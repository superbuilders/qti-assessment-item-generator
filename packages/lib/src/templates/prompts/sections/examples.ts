import { existsSync, readFileSync } from "node:fs"
import * as path from "node:path"
import { resolveLibPath } from "@/internal/paths"

const POSITIVE_EXAMPLE_CACHE = new Map<string, string>()
const NEGATIVE_EXAMPLE_CACHE = new Map<string, string>()

const POSITIVE_ROOT = resolveLibPath("templates/prompts/examples/positive")
const POSITIVE_TEMPLATES_DIR = path.join(POSITIVE_ROOT, "templates")
const POSITIVE_NOTES_DIR = path.join(POSITIVE_ROOT, "notes")

const NEGATIVE_ROOT = resolveLibPath("templates/prompts/examples/negative")
const NEGATIVE_TEMPLATES_DIR = path.join(NEGATIVE_ROOT, "templates")
const NEGATIVE_NOTES_DIR = path.join(NEGATIVE_ROOT, "notes")

export function createPositiveExampleSection(names: readonly string[]): string {
	if (names.length === 0) return ""
	const rendered = names
		.map((name) => loadPositiveExample(name))
		.filter((block) => block.length > 0)
		.join("\n\n")
	if (rendered.length === 0) return ""
	return `### POSITIVE_EXAMPLES
<examples type="positive">
${rendered}
</examples>`
}

export function createNegativeExampleSection(names: readonly string[]): string {
	if (names.length === 0) return ""
	const rendered = names
		.map((name) => loadNegativeExample(name))
		.filter((block) => block.length > 0)
		.join("\n\n")
	if (rendered.length === 0) return ""
	return `### NEGATIVE_EXAMPLES
<examples type="negative">
${rendered}
</examples>`
}

function loadPositiveExample(name: string): string {
	const cached = POSITIVE_EXAMPLE_CACHE.get(name)
	if (cached) return cached

	const codePath = path.join(POSITIVE_TEMPLATES_DIR, `${name}.ts`)
	if (!existsSync(codePath)) {
		return ""
	}

	const notesPath = path.join(POSITIVE_NOTES_DIR, `${name}.md`)
	const code = readFileSync(codePath, "utf-8")
	const notes = existsSync(notesPath)
		? readFileSync(notesPath, "utf-8").trim()
		: ""

	const rendered = `<example kind="positive" name="${name}" source="${codePath}">
<code>
${code}
</code>
${notes ? `<notes>\n${notes}\n</notes>` : "<notes />"}
</example>`
	POSITIVE_EXAMPLE_CACHE.set(name, rendered)
	return rendered
}

function loadNegativeExample(name: string): string {
	const existing = NEGATIVE_EXAMPLE_CACHE.get(name)
	if (existing) return existing

	const codePath = path.join(NEGATIVE_TEMPLATES_DIR, `${name}.ts`)
	if (!existsSync(codePath)) {
		return ""
	}
	const notesPath = path.join(NEGATIVE_NOTES_DIR, `${name}.md`)
	const code = readFileSync(codePath, "utf-8")
	const notes = existsSync(notesPath)
		? readFileSync(notesPath, "utf-8").trim()
		: ""
	const rendered = `<example kind="negative" name="${name}" source="${codePath}">
<code>
${code}
</code>
${notes ? `<notes>\n${notes}\n</notes>` : "<notes />"}
</example>`
	NEGATIVE_EXAMPLE_CACHE.set(name, rendered)
	return rendered
}
