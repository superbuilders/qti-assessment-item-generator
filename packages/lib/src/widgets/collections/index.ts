import { allWidgetsCollection } from "@/widgets/collections/all"
import { fourthGradeMathCollection } from "@/widgets/collections/fourth-grade-math"
import { mathCoreCollection } from "@/widgets/collections/math-core"
import { scienceCollection } from "@/widgets/collections/science"
import { simpleVisualCollection } from "@/widgets/collections/simple-visual"
import { teksMath4Collection } from "@/widgets/collections/teks-math-4"

export const widgetCollections: {
	readonly all: typeof allWidgetsCollection
	readonly science: typeof scienceCollection
	readonly "simple-visual": typeof simpleVisualCollection
	readonly "fourth-grade-math": typeof fourthGradeMathCollection
	readonly "teks-math-4": typeof teksMath4Collection
	readonly "math-core": typeof mathCoreCollection
} = {
	all: allWidgetsCollection,
	science: scienceCollection,
	"simple-visual": simpleVisualCollection,
	"fourth-grade-math": fourthGradeMathCollection,
	"teks-math-4": teksMath4Collection,
	"math-core": mathCoreCollection
}

export type { WidgetCollection } from "@/widgets/collections/types"
