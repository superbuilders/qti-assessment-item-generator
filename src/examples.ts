import type { AssessmentItemInput } from "@/core/item"
import type { allWidgetsCollection } from "@/widgets/collections/all"
import type { WidgetTypeTupleFrom } from "@/widgets/collections/types"

export const allExamples: AssessmentItemInput<WidgetTypeTupleFrom<typeof allWidgetsCollection>>[] = []
