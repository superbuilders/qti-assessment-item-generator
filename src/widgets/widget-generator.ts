import * as errors from "@superbuilders/errors"
import * as logger from "@superbuilders/slog"
import { generateThreeDIntersectionDiagram } from "./generators/3d-intersection-diagram"
import { generateAdditionWithRegrouping } from "./generators/addition-with-regrouping"
import { generateAngleDiagram } from "./generators/angle-diagram"
import { generateAngleTypeDiagram } from "./generators/angle-type-diagram"
import { generateAreaGraph } from "./generators/area-graph"
import { generateAreaModelMultiplication } from "./generators/area-model-multiplication"
import { generateBarChart } from "./generators/bar-chart"
import { generateBoxGrid } from "./generators/box-grid"
import { generateBoxPlot } from "./generators/box-plot"
import { generateCircleAngleDiagram } from "./generators/circle-angle-diagram"
import { generateCircleDiagram } from "./generators/circle-diagram"
import { generateClockDiagram } from "./generators/clock-diagram"
import { generateCompositeShapeDiagram } from "./generators/composite-shape-diagram"
import { generateConceptualGraph } from "./generators/conceptual-graph"
import { generateConstraintGeometryDiagram } from "./generators/constraint-geometry-diagram"
import { generateCoordinatePlaneComprehensive } from "./generators/coordinate-plane-comprehensive"
import { generateCustomPolygonDiagram } from "./generators/custom-polygon-diagram"
import { generateDiscreteObjectRatioDiagram } from "./generators/discrete-object-ratio-diagram"
import { generateDivergentBarChart } from "./generators/divergent-bar-chart"
import { generateDivisionAreaDiagram } from "./generators/division-area-diagram"
import { generateDivisionModelDiagram } from "./generators/division-model-diagram"
import { generateDotPlot } from "./generators/dot-plot"
import { generateDoubleNumberLine } from "./generators/double-number-line"
import { generateEmojiImage } from "./generators/emoji-image"
import { generateEquivalentFractionModel } from "./generators/equivalent-fraction-model"
import { generateFactorizationDiagram } from "./generators/factorization-diagram"
import { generateFigureComparisonDiagram } from "./generators/figure-comparison-diagram"
import { generateFractionFrequencyPlot } from "./generators/fraction-frequency-plot"
import { generateFractionMultiplicationModel } from "./generators/fraction-multiplication-model"
import { generateFractionSumDiagram } from "./generators/fraction-sum-diagram"
import { generateFractionModelDiagram } from "./generators/fractional-model-diagram"
import { generateFreeBodyDiagram } from "./generators/free-body-diagram"
import { generateFunctionPlotGraph } from "./generators/function-plot-graph"
import { generateGeometricPrimitiveDiagram } from "./generators/geometric-primitive-diagram"
import { generateGeometricSolidDiagram } from "./generators/geometric-solid-diagram"
import { generateHangerDiagram } from "./generators/hanger-diagram"
import { generateHistogram } from "./generators/histogram"
import { generateInequalityNumberLine } from "./generators/inequality-number-line"
import { generateIntegerChipLegend } from "./generators/integer-chip-legend"
import { generateIntegerChipModel } from "./generators/integer-chip-model"
import { generateKeelingCurve } from "./generators/keeling-curve"
import { generateLabeledRectangleDiagram } from "./generators/labeled-rectangle-diagram"
import { generateLineDiagram } from "./generators/line-diagram"
import { generateLineEquationGraph } from "./generators/line-equation-graph"
import { generateLineGraph } from "./generators/line-graph"
import { generateMarbleDiagram } from "./generators/marble-diagram"
import { generateNPolygon } from "./generators/n-polygon"
import { generateNumberLine } from "./generators/number-line"
// removed: number-line-for-opposites
import { generateNumberLineWithAction } from "./generators/number-line-with-action"
import { generateNumberLineWithFractionGroups } from "./generators/number-line-with-fraction-groups"
import { generateNumberSetDiagram } from "./generators/number-set-diagram"
import { generateParabolaGraph } from "./generators/parabola-graph"
import { generateParallelogramTrapezoidDiagram } from "./generators/parallelogram-trapezoid-diagram"
import { generatePartitionedShape } from "./generators/partitioned-shape"
import { generatePatternDiagram } from "./generators/pattern-diagram"
import { generatePentagonIntersectionDiagram } from "./generators/pentagon-intersection-diagram"
import { generatePeriodicTable } from "./generators/periodic-table"
import { generatePESSpectrum } from "./generators/pes-spectrum"
import { generatePieChart } from "./generators/pi-chart"
import { generatePictograph } from "./generators/pictograph"
import { generatePointPlotGraph } from "./generators/point-plot-graph"
import { generatePolygonGraph } from "./generators/polygon-graph"
import { generatePolyhedronDiagram } from "./generators/polyhedron-diagram"
import { generatePolyhedronNetDiagram } from "./generators/polyhedron-net-diagram"
import { generatePopulationBarChart } from "./generators/population-bar-chart"
import { generatePopulationChangeEventGraph } from "./generators/population-change-event-graph"
import { generateProbabilitySpinner } from "./generators/probability-spinner"
import { generateProtractorAngleDiagram } from "./generators/protractor-angle-diagram"
import { generatePythagoreanProofDiagram } from "./generators/pythagorean-proof-diagram"
import { generateQuadrantDiagram } from "./generators/quadrant-diagram"
import { generateQuantityFractionalDiagram } from "./generators/quantity-fractional-diagram"
import { generateRadiallyConstrainedAngleDiagram } from "./generators/radially-constrained-angle-diagram"
import { generateRatioBoxDiagram } from "./generators/ratio-box-diagram"
import { generateRectangularFrameDiagram } from "./generators/rectangular-frame-diagram"
import { generateScaleCopiesSlider } from "./generators/scale-copies-slider"
import { generateScatterPlot } from "./generators/scatter-plot"
import { generateShapeTransformationGraph } from "./generators/shape-transformation-graph"
import { generateSimpleArrow } from "./generators/simple-arrow"
import { generateSinCosineWidget } from "./generators/sin-cosine-widget"
import { generateStackedItemsDiagram } from "./generators/stacked-items-diagram"
import { generateStickPlot } from "./generators/stick-plot"
import { generateSubtractionWithRegrouping } from "./generators/subtraction-with-regrouping"
import { generateSymmetryDiagram } from "./generators/symmetry-diagram"
import { generateTapeDiagram } from "./generators/tape-diagram"
import { generateTransformationDiagram } from "./generators/transformation-diagram"
import { generateTransversalAngleDiagram } from "./generators/transversal-angle-diagram"
import { generateTreeDiagram } from "./generators/tree-diagram"
import { generateTriangleDiagram } from "./generators/triangle-diagram"
import { generateUnitBlockDiagram } from "./generators/unit-block-diagram"
import { generateUrlImage } from "./generators/url-image"
import { generateVectorDiagram } from "./generators/vector-diagram"
import { generateVennDiagram } from "./generators/venn-diagram"
import { generateVerticalArithmeticSetup } from "./generators/vertical-arithmetic-setup"
import { generateWheelDiagram } from "./generators/wheel-diagram"
import { type WidgetInput, WidgetSchema } from "./registry"

export async function generateWidget(widgetInput: WidgetInput): Promise<string> {
	const parsed = WidgetSchema.safeParse(widgetInput)
	if (!parsed.success) {
		logger.error("widget validation failed", { error: parsed.error })
		throw errors.wrap(parsed.error, "widget validation")
	}
	const widget = parsed.data
	switch (widget.type) {
		case "quadrantDiagram":
			return await generateQuadrantDiagram(widget)
		case "clockDiagram":
			return await generateClockDiagram(widget)
		case "wheelDiagram":
			return await generateWheelDiagram(widget)
		case "labeledRectangleDiagram":
			return await generateLabeledRectangleDiagram(widget)
		case "customPolygonDiagram":
			return await generateCustomPolygonDiagram(widget)
		case "lineDiagram":
			return await generateLineDiagram(widget)
		case "geometricPrimitiveDiagram":
			return await generateGeometricPrimitiveDiagram(widget)
		case "constraintGeometryDiagram":
			return await generateConstraintGeometryDiagram(widget)
		case "areaGraph":
			return await generateAreaGraph(widget)
		case "areaModelMultiplication":
			return await generateAreaModelMultiplication(widget)
		case "threeDIntersectionDiagram":
			return await generateThreeDIntersectionDiagram(widget)
		case "angleDiagram":
			return await generateAngleDiagram(widget)
		case "angleTypeDiagram":
			return await generateAngleTypeDiagram(widget)
		case "conceptualGraph":
			return await generateConceptualGraph(widget)
		case "barChart":
			return await generateBarChart(widget)
		case "boxGrid":
			return await generateBoxGrid(widget)
		case "boxPlot":
			return await generateBoxPlot(widget)
		case "circleDiagram":
			return await generateCircleDiagram(widget)
		case "circleAngleDiagram":
			return await generateCircleAngleDiagram(widget)
		case "fractionModelDiagram":
			return await generateFractionModelDiagram(widget)
		case "compositeShapeDiagram":
			return await generateCompositeShapeDiagram(widget)
		case "coordinatePlane":
			return await generateCoordinatePlaneComprehensive(widget)
		case "divergentBarChart":
			return await generateDivergentBarChart(widget)
		case "functionPlotGraph":
			return await generateFunctionPlotGraph(widget)
		case "keelingCurve":
			return await generateKeelingCurve(widget)
		case "lineEquationGraph":
			return await generateLineEquationGraph(widget)
		case "lineGraph":
			return await generateLineGraph(widget)
		case "pointPlotGraph":
			return await generatePointPlotGraph(widget)
		case "polygonGraph":
			return await generatePolygonGraph(widget)
		case "shapeTransformationGraph":
			return await generateShapeTransformationGraph(widget)
		case "discreteObjectRatioDiagram":
			return await generateDiscreteObjectRatioDiagram(widget)
		case "dotPlot":
			return await generateDotPlot(widget)
		case "doubleNumberLine":
			return await generateDoubleNumberLine(widget)
		case "parabolaGraph":
			return await generateParabolaGraph(widget)
		case "populationBarChart":
			return await generatePopulationBarChart(widget)
		case "populationChangeEventGraph":
			return await generatePopulationChangeEventGraph(widget)
		case "geometricSolidDiagram":
			return await generateGeometricSolidDiagram(widget)
		case "hangerDiagram":
			return await generateHangerDiagram(widget)
		case "histogram":
			return await generateHistogram(widget)
		case "nPolygon":
			return await generateNPolygon(widget)
		case "patternDiagram":
			return await generatePatternDiagram(widget)
		case "inequalityNumberLine":
			return await generateInequalityNumberLine(widget)
		case "numberLine":
			return await generateNumberLine(widget)
		case "numberLineWithAction":
			return await generateNumberLineWithAction(widget)
		case "numberLineWithFractionGroups":
			return await generateNumberLineWithFractionGroups(widget)
		case "numberSetDiagram":
			return await generateNumberSetDiagram(widget)
		case "partitionedShape":
			return await generatePartitionedShape(widget)
		case "pentagonIntersectionDiagram":
			return await generatePentagonIntersectionDiagram(widget)
		case "pictograph":
			return await generatePictograph(widget)
		case "integerChipModel":
			return await generateIntegerChipModel(widget)
		case "integerChipLegend":
			return await generateIntegerChipLegend(widget)
		case "polyhedronDiagram":
			return await generatePolyhedronDiagram(widget)
		case "probabilitySpinner":
			return await generateProbabilitySpinner(widget)
		case "protractorAngleDiagram":
			return await generateProtractorAngleDiagram(widget)
		case "radiallyConstrainedAngleDiagram":
			return await generateRadiallyConstrainedAngleDiagram(widget)
		case "polyhedronNetDiagram":
			return await generatePolyhedronNetDiagram(widget)
		case "pythagoreanProofDiagram":
			return await generatePythagoreanProofDiagram(widget)
		case "ratioBoxDiagram":
			return await generateRatioBoxDiagram(widget)
		case "rectangularFrameDiagram":
			return await generateRectangularFrameDiagram(widget)
		case "scaleCopiesSlider":
			return await generateScaleCopiesSlider(widget)
		case "scatterPlot":
			return await generateScatterPlot(widget)
		case "fractionMultiplicationModel":
			return await generateFractionMultiplicationModel(widget)
		case "stackedItemsDiagram":
			return await generateStackedItemsDiagram(widget)
		case "tapeDiagram":
			return await generateTapeDiagram(widget)
		case "transformationDiagram":
			return await generateTransformationDiagram(widget)
		case "transversalAngleDiagram":
			return await generateTransversalAngleDiagram(widget)
		case "treeDiagram":
			return await generateTreeDiagram(widget)
		case "triangleDiagram":
			return await generateTriangleDiagram(widget)
		case "unitBlockDiagram":
			return await generateUnitBlockDiagram(widget)
		case "periodicTable":
			return await generatePeriodicTable(widget)
		case "vennDiagram":
			return await generateVennDiagram(widget)
		case "verticalArithmeticSetup":
			return await generateVerticalArithmeticSetup(widget)
		case "emojiImage":
			return await generateEmojiImage(widget)
		case "figureComparisonDiagram":
			return await generateFigureComparisonDiagram(widget)
		case "fractionSumDiagram":
			return await generateFractionSumDiagram(widget)
		case "parallelogramTrapezoidDiagram":
			return await generateParallelogramTrapezoidDiagram(widget)
		case "urlImage":
			return await generateUrlImage(widget)
		case "pieChart":
			return await generatePieChart(widget)
		case "subtractionWithRegrouping":
			return await generateSubtractionWithRegrouping(widget)
		case "fractionFrequencyPlot":
			return await generateFractionFrequencyPlot(widget)
		case "divisionAreaDiagram":
			return await generateDivisionAreaDiagram(widget)
		case "divisionModelDiagram":
			return await generateDivisionModelDiagram(widget)
		case "factorizationDiagram":
			return await generateFactorizationDiagram(widget)
		case "freeBodyDiagram":
			return await generateFreeBodyDiagram(widget)
		case "equivalentFractionModel":
			return await generateEquivalentFractionModel(widget)
		case "additionWithRegrouping":
			return await generateAdditionWithRegrouping(widget)
		case "quantityFractionalDiagram":
			return await generateQuantityFractionalDiagram(widget)
		case "simpleArrow":
			return await generateSimpleArrow(widget)
		case "sinCosineWidget":
			return await generateSinCosineWidget(widget)
		case "vectorDiagram":
			return await generateVectorDiagram(widget)
		case "pesSpectrum":
			return await generatePESSpectrum(widget)
		case "stickPlot":
			return await generateStickPlot(widget)
		case "marbleDiagram":
			return await generateMarbleDiagram(widget)
		case "symmetryDiagram":
			return await generateSymmetryDiagram(widget)
		default:
			logger.error("unknown widget type", { widget })
			throw errors.new(`Unknown widget type: ${JSON.stringify(widget)}`)
	}
}
