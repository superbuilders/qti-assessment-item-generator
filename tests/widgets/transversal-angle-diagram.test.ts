import { describe, expect, test } from "bun:test"
import * as errors from "@superbuilders/errors"
import { transversalAngleDiagramExamples } from "../../examples/transversal-angle-diagram"
import type { Widget } from "../../src/widgets/registry"
import { generateWidget } from "../../src/widgets/widget-generator"

describe("Widget: transversal-angle-diagram", () => {
	const examples = transversalAngleDiagramExamples as unknown as Widget[]

	examples.forEach((props, index) => {
		test(`should produce consistent output for example #${index + 1}`, async () => {
			const result = await errors.try(generateWidget(props))
			if (result.error) throw result.error
			expect(result.data).toMatchSnapshot()
		})
	})

	test("throws error for transversal parallel to a main line", async () => {
		const result = await errors.try(generateWidget({
			type: "transversalAngleDiagram",
			width: 400,
			height: 400,
			line1Angle: 45,
			line2Angle: 0,
			transversalAngle: 45, // Same as line1 -> invalid
			intersections: {
				line1: {
					label: "P",
					angles: [
						{ span: "mainToTransversal", label: "1", color: "#ff6b6b", fromRay: { dir: "-" }, toRay: { dir: "+" } }
					]
				},
				line2: {
					label: "Q",
					angles: [
						{ span: "mainToTransversal", label: "2", color: "#4ecdc4", fromRay: { dir: "-" }, toRay: { dir: "+" } }
					]
				}
			},
			pointLabels: {
				line1: ["A", "B"],
				line2: ["C", "D"],
				transversal: ["E", "F"]
			}
		} as Widget))
		expect(result.error).toBeTruthy()
	})

	test("throws error for missing line1 intersection", async () => {
		const result = await errors.try(generateWidget({
			type: "transversalAngleDiagram",
			width: 400,
			height: 400,
			line1Angle: 0,
			line2Angle: 0,
			transversalAngle: 45,
			intersections: {
				line2: {
					label: "P",
					angles: [
						{ span: "mainToTransversal", label: "1", color: "#ff6b6b", fromRay: { dir: "-" }, toRay: { dir: "+" } }
					]
				}
			} as any
		} as Widget))
		expect(result.error).toBeTruthy()
	})

	test("throws error for duplicate intersection labels", async () => {
		const result = await errors.try(generateWidget({
			type: "transversalAngleDiagram",
			width: 400,
			height: 400,
			line1Angle: 0,
			line2Angle: 0,
			transversalAngle: 45,
			intersections: {
				line1: { label: "P", angles: [ { span: "mainToTransversal", label: "1", color: "#ff6b6b", fromRay: { dir: "-" }, toRay: { dir: "+" } } ] },
				line2: { label: "P", angles: [ { span: "mainToTransversal", label: "2", color: "#4ecdc4", fromRay: { dir: "-" }, toRay: { dir: "+" } } ] }
			}
		} as Widget))
		expect(result.error).toBeTruthy()
	})

	test("throws error for angle not spanning main and transversal (cannot construct with schema)", async () => {
		// This case is no longer buildable via schema, but we keep a sanity check using any-cast
		const badAngles = [{ span: "mainToTransversal", label: "1", color: "#ff6b6b", fromRay: { dir: "-" }, toRay: { dir: "+" } }]
		const result = await errors.try(generateWidget({
			type: "transversalAngleDiagram",
			width: 400,
			height: 400,
			line1Angle: 0,
			line2Angle: 0,
			transversalAngle: 45,
			intersections: {
				line1: { label: "P", angles: badAngles as any },
				line2: { label: "Q", angles: badAngles as any }
			}
		} as Widget))
		// If schema allows construction, generator would still succeed. This ensures no throw.
		expect(result.error).toBeFalsy()
	})

	test("should still error when angle arrays are missing (invalid structure)", async () => {
		const invalid = await errors.try(generateWidget({
			type: "transversalAngleDiagram",
			width: 400,
			height: 400,
			line1Angle: 0,
			line2Angle: 0,
			transversalAngle: 45,
			intersections: {
				line1: { label: "P" } as any,
				line2: { label: "Q", angles: [ { span: "mainToTransversal", label: "1", color: "#ff6b6b", fromRay: { dir: "-" }, toRay: { dir: "+" } } ] }
			}
		} as Widget))
		expect(invalid.error).toBeTruthy()
	})
})
