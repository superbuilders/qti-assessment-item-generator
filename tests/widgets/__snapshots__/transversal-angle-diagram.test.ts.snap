// Bun Snapshot v1, https://bun.sh/docs/test/snapshots

exports[`Widget: transversal-angle-diagram should produce consistent output for example #1 1`] = `"<svg width="540" height="490" viewBox="-20 -20 540 490" xmlns="http://www.w3.org/2000/svg" font-family="sans-serif"><path d="M 230.76136693793114 277.85770991920737 L 219.21818710068982 309.57233587073176 A 33.75 33.75 0 0 1 197.01136693793114 277.85770991920737 Z" fill="#5DADE2" stroke="none"/><path d="M 230.76136693793114 277.85770991920737 L 219.21818710068982 309.57233587073176 A 33.75 33.75 0 0 0 264.51136693793114 277.85770991920737 Z" fill="#F5B041" stroke="none"/><path d="M 269.23863306206886 172.14229008079266 L 257.69545322482753 203.8569160323171 A 33.75 33.75 0 0 1 235.48863306206886 172.14229008079266 Z" fill="#F7DC6F" stroke="none"/><path d="M 269.23863306206886 172.14229008079266 L 257.69545322482753 203.8569160323171 A 33.75 33.75 0 0 0 302.98863306206886 172.14229008079266 Z" fill="#58D68D" stroke="none"/><path d="M 269.23863306206886 172.14229008079266 L 302.98863306206886 172.14229008079266 A 33.75 33.75 0 0 0 280.7818128993102 140.42766412926827 Z" fill="#AF601A" stroke="none"/><line x1="30.761366937931143" y1="277.85770991920737" x2="430.76136693793114" y2="277.85770991920737" stroke="#000000" stroke-width="2"/><line x1="69.23863306206886" y1="172.14229008079266" x2="469.23863306206886" y2="172.14229008079266" stroke="#000000" stroke-width="2"/><line x1="181.59597133486625" y1="412.93852415718163" x2="318.40402866513375" y2="37.06147584281834" stroke="#000000" stroke-width="2"/><text x="230.76136693793114" y="263.85770991920737" font-size="18" text-anchor="middle" dominant-baseline="middle" font-weight="700" fill="#000000">M</text><text x="269.23863306206886" y="158.14229008079266" font-size="18" text-anchor="middle" dominant-baseline="middle" font-weight="700" fill="#000000">N</text><text x="191.6468568231318" y="305.2459847549698" font-size="16" text-anchor="middle" dominant-baseline="middle" font-weight="700" fill="#000000">1</text><text x="258.1496417736936" y="316.97222003400674" font-size="16" text-anchor="middle" dominant-baseline="middle" font-weight="700" fill="#000000">2</text><text x="230.1241229472695" y="199.53056491655514" font-size="16" text-anchor="middle" dominant-baseline="middle" font-weight="700" fill="#000000">3</text><text x="296.6269078978313" y="211.256800195592" font-size="16" text-anchor="middle" dominant-baseline="middle" font-weight="700" fill="#000000">4</text><text x="308.35314317686823" y="144.75401524503022" font-size="16" text-anchor="middle" dominant-baseline="middle" font-weight="700" fill="#000000">5</text><circle cx="388.26136693793114" cy="277.85770991920737" r="5" fill="#000000"/><text x="403.8177161240352" y="293.4140591053114" font-size="20" text-anchor="middle" dominant-baseline="middle" font-weight="700" fill="#000000">D</text><circle cx="73.26136693793114" cy="277.85770991920737" r="5" fill="#000000"/><text x="88.81771612403519" y="293.4140591053114" font-size="20" text-anchor="middle" dominant-baseline="middle" font-weight="700" fill="#000000">C</text><circle cx="176.8931943641383" cy="425.859297692988" r="5" fill="#000000"/><text x="198.8931943641383" y="425.859297692988" font-size="20" text-anchor="middle" dominant-baseline="middle" font-weight="700" fill="#000000">E</text><circle cx="426.73863306206886" cy="172.14229008079266" r="5" fill="#000000"/><text x="442.2949822481729" y="187.6986392668967" font-size="20" text-anchor="middle" dominant-baseline="middle" font-weight="700" fill="#000000">B</text><circle cx="111.73863306206886" cy="172.1422900807927" r="5" fill="#000000"/><text x="127.2949822481729" y="187.69863926689675" font-size="20" text-anchor="middle" dominant-baseline="middle" font-weight="700" fill="#000000">A</text><circle cx="323.1068056358617" cy="24.140702307012106" r="5" fill="#000000"/><text x="345.1068056358617" y="24.140702307012106" font-size="20" text-anchor="middle" dominant-baseline="middle" font-weight="700" fill="#000000">F</text></svg>"`;

exports[`Widget: transversal-angle-diagram should produce consistent output for example #2 1`] = `"<svg width="540" height="490" viewBox="-20 -20 540 490" xmlns="http://www.w3.org/2000/svg" font-family="sans-serif"><path d="M 201.28607103712534 253.125 L 172.05771365940052 270 A 33.75 33.75 0 0 0 233.00069698864974 264.6681798372413 Z" fill="#5DADE2" stroke="none"/><path d="M 201.28607103712534 253.125 L 172.05771365940052 270 A 33.75 33.75 0 0 1 169.57144508560094 241.58182016275867 Z" fill="#F5B041" stroke="none"/><path d="M 201.28607103712534 253.125 L 169.57144508560094 241.58182016275867 A 33.75 33.75 0 0 1 230.51442841485016 236.25 Z" fill="#C39BD3" stroke="none"/><path d="M 298.71392896287466 196.875 L 269.48557158514984 213.75 A 33.75 33.75 0 0 0 331.9511906270367 202.7356259962589 Z" fill="#F7DC6F" stroke="none"/><path d="M 298.71392896287466 196.875 L 331.9511906270367 202.7356259962589 A 33.75 33.75 0 0 0 327.9422863405995 180 Z" fill="#58D68D" stroke="none"/><path d="M 298.71392896287466 196.875 L 265.47666729871264 191.01437400374112 A 33.75 33.75 0 0 1 327.9422863405995 180 Z" fill="#AF601A" stroke="none"/><line x1="13.34754687994365" y1="184.72097133486625" x2="389.224595194307" y2="321.52902866513375" stroke="#000000" stroke-width="2"/><line x1="101.75237836043306" y1="162.14536446661393" x2="495.67547956531627" y2="231.60463553338607" stroke="#000000" stroke-width="2"/><line x1="76.79491924311225" y1="325" x2="423.20508075688775" y2="125.00000000000001" stroke="#000000" stroke-width="2"/><text x="201.28607103712534" y="239.125" font-size="18" text-anchor="middle" dominant-baseline="middle" font-weight="700" fill="#000000">G</text><text x="298.71392896287466" y="182.875" font-size="18" text-anchor="middle" dominant-baseline="middle" font-weight="700" fill="#000000">H</text><text x="205.44775775332602" y="300.6932968338808" font-size="16" text-anchor="middle" dominant-baseline="middle" font-weight="700" fill="#000000">1</text><text x="153.7177742032445" y="257.2866867162007" font-size="16" text-anchor="middle" dominant-baseline="middle" font-weight="700" fill="#000000">2</text><text x="197.12438432092466" y="205.55670316611915" font-size="16" text-anchor="middle" dominant-baseline="middle" font-weight="700" fill="#000000">6</text><text x="307.0056294464706" y="243.89957020633292" font-size="16" text-anchor="middle" dominant-baseline="middle" font-weight="700" fill="#000000">5</text><text x="345.7384991692076" y="188.58329951640408" font-size="16" text-anchor="middle" dominant-baseline="middle" font-weight="700" fill="#000000">4</text><text x="290.4222284792787" y="149.85042979366708" font-size="16" text-anchor="middle" dominant-baseline="middle" font-weight="700" fill="#000000">3</text><circle cx="349.2876588109059" cy="306.99317257379283" r="5" fill="#000000"/><text x="349.2876588109059" y="328.99317257379283" font-size="20" text-anchor="middle" dominant-baseline="middle" font-weight="700" fill="#000000">C</text><circle cx="53.284483263344754" cy="199.2568274262072" r="5" fill="#000000"/><text x="53.284483263344754" y="221.2568274262072" font-size="20" text-anchor="middle" dominant-baseline="middle" font-weight="700" fill="#000000">D</text><circle cx="64.88706994107625" cy="331.875" r="5" fill="#000000"/><text x="80.44341912718029" y="347.43134918610406" font-size="20" text-anchor="middle" dominant-baseline="middle" font-weight="700" fill="#000000">E</text><circle cx="453.8211500622974" cy="224.22458798254152" r="5" fill="#000000"/><text x="476.44856706026695" y="246.85200498051103" font-size="20" text-anchor="middle" dominant-baseline="middle" font-weight="700" fill="#000000">A</text><circle cx="143.60670786345187" cy="169.5254120174585" r="5" fill="#000000"/><text x="166.2341248614214" y="192.15282901542804" font-size="20" text-anchor="middle" dominant-baseline="middle" font-weight="700" fill="#000000">B</text><circle cx="435.1129300589238" cy="118.12500000000001" r="5" fill="#000000"/><text x="457.1129300589238" y="118.12500000000001" font-size="20" text-anchor="middle" dominant-baseline="middle" font-weight="700" fill="#000000">F</text></svg>"`;
